scalyr_agent/builtin_monitors/tests/journald_monitor_test.py

```diff
@@ -0,0 +1,254 @@
+
+    @mock.patch(
+        "scalyr_agent.builtin_monitors.journald_monitor.verify_systemd_library_is_available",
+        side_effect=empty,
+    )
+    def test_write_to_disk(self, verify):
+        fake_journal = {
+            "_AUDIT_LOGINUID": 1000,
+            "_CAP_EFFECTIVE": "0",
+            "_SELINUX_CONTEXT": "unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023",
+            "_GID": 1000,
+            "CODE_LINE": 1,
+            "_HOSTNAME": "...",
+            "_SYSTEMD_SESSION": 52,
+            "_SYSTEMD_OWNER_UID": 1000,
+            "MESSAGE": "testing 1,2,3",
+            # '__MONOTONIC_TIMESTAMP':
+            # journal.Monotonic(timestamp=datetime.timedelta(2, 76200, 811585), bootid=UUID('958b7e26-df4c-453a-a0f9-a8406cb508f2')),
+            "SYSLOG_IDENTIFIER": "python3",
+            "_UID": 1000,
+            "_EXE": "/usr/bin/python3",
+            "_PID": 7733,
+            "_COMM": "...",
+            "CODE_FUNC": "<module>",
+            "CODE_FILE": "<doctest journal.rst[4]>",
+            "_SOURCE_REALTIME_TIMESTAMP": datetime.datetime(
+                2015, 9, 5, 13, 17, 4, 944355
+            ),
+            # '__CURSOR': 's=...',
+            "_BOOT_ID": UUID("958b7e26-df4c-453a-a0f9-a8406cb508f2"),
+            "_CMDLINE": "/usr/bin/python3 ...",
+            "_MACHINE_ID": UUID("263bb31e-3e13-4062-9bdb-f1f4518999d2"),
+            "_SYSTEMD_SLICE": "user-1000.slice",
+            "_AUDIT_SESSION": 52,
+            "__REALTIME_TIMESTAMP": datetime.datetime(2015, 9, 5, 13, 17, 4, 945110),
+            "_SYSTEMD_UNIT": "session-52.scope",
+            "_SYSTEMD_CGROUP": "/user.slice/user-1000.slice/session-52.scope",
+            "_TRANSPORT": "journal",
+        }
+
+        def fake_pending_entries(self):
+            self._journal = [fake_journal]
+            return True
+
+        with mock.patch.object(
+            JournaldMonitor, "_has_pending_entries", fake_pending_entries
+        ):
+            journal_directory = tempfile.mkdtemp(suffix="journal")
+            fake_clock = FakeClock()
+            manager_poll_interval = 30
+            manager, _ = ScalyrTestUtils.create_test_monitors_manager(
+                config_monitors=[
+                    {
+                        "module": "scalyr_agent.builtin_monitors.journald_monitor",
+                        "journal_path": journal_directory,
+                    }
+                ],
+                extra_toplevel_config={
+                    "user_agent_refresh_interval": manager_poll_interval,
+                    "agent_log_path": journal_directory,
+                },
+                null_logger=True,
+                fake_clock=fake_clock,
+            )
+            monitor = manager.monitors[0]
+            monitor.log_manager.set_log_watcher(LogWatcher())
+
+            monitor.gather_sample()
+
+            self.assertLogFileContainsLineRegex(
+                "....\\-..\\-.. ..\\:..\\:..\\.....",
```

Comment:
Why do you need this regex?  Is it because the timestamp is non-deterministic?  It looks like you are trying to use a timestamp.  It would be better if we could just write out the whole line that should be produced by journald and check that explicitly here.

Reply:
Testing the exact line also guarantees that all of these extra fields are on the same line and that there are no fields other than the ones listed included in the output.

Reply:
Timestamp is generated from the actual time, I can mock that out somehow I'm sure.